splash/browser_tab.py

```diff
@@ -503,6 +511,64 @@ def runjs(self, js_source, handle_errors=True):
         js_source = "%s;undefined" % js_source
         self.evaljs(js_source, handle_errors=handle_errors)
 
+    def wait_for_resume(self, js_source, callback, errback, timeout):
+        """
+        Run some Javascript asynchronously.
+
+        The JavaScript must contain a method called `main()` that accepts
+        one argument. The first argument will be an object with `resume()`
+        and `error()` methods. The code _must_ call one of these functions
+        before the timeout or else it will be canceled.
+
+        Note: this cleans up the JavaScript global variable that it creates,
+        but QT seems to notice when a JS GV is deleted and it destroys the
+        underlying C++ object. Therefore, we can only delete the JS GV _after_
+        the user's code has called us back. This should change in QT5, since
+        it will then be possible to specify a different object ownership
+        policy when calling addToJavaScriptWindowObject().
+        """
+
+        frame = self.web_page.mainFrame()
+        script_text = js_source.replace('\n', ' \\\n').replace('\"', '\\"')
+        callback_proxy = OneShotCallbackProxy(self, callback, errback, timeout)
+        frame.addToJavaScriptWindowObject(callback_proxy.name, callback_proxy)
+
+        wrapped = """
+        (function () {
+            try {
+                eval("%(script_text)s");
+            } catch (err) {
+                var main = function (splash) {
+                    splash.error(err);
```

Comment:
hmm, I used err.toString() in `_WrappedJavascriptFunction.__call__` - does it produce a different result? If no, it makes sense to remove .toString() in other places.